vapi/appliance/simulator/simulator.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/vmware/govmomi/simulator"
+	"github.com/vmware/govmomi/vapi/appliance/access/consolecli"
+	"github.com/vmware/govmomi/vapi/appliance/access/dcui"
+	"github.com/vmware/govmomi/vapi/appliance/access/shell"
+	"github.com/vmware/govmomi/vapi/appliance/access/ssh"
+	vapi "github.com/vmware/govmomi/vapi/simulator"
+)
+
+func init() {
+	simulator.RegisterEndpoint(func(s *simulator.Service, r *simulator.Registry) {
+		New(s.Listen).Register(s, r)
+	})
+}
+
+// Handler implements the Appliance API simulator
+type Handler struct {
+	URL        *url.URL
+	consolecli struct {
+		access consolecli.Access
+		sync.RWMutex
+	}
+	dcui struct {
+		access dcui.Access
+		sync.RWMutex
+	}
+	ssh struct {
+		access ssh.Access
+		sync.RWMutex
+	}
+	shell struct {
+		access shell.Access
+		sync.RWMutex
+	}
+}
+
+// New creates a Handler instance
+func New(u *url.URL) *Handler {
+	return &Handler{
+		URL: nil,
+		consolecli: struct {
+			access consolecli.Access
+			sync.RWMutex
+		}{},
+		dcui: struct {
+			access dcui.Access
+			sync.RWMutex
+		}{},
+		ssh: struct {
+			access ssh.Access
+			sync.RWMutex
+		}{},
+		shell: struct {
+			access shell.Access
+			sync.RWMutex
+		}{},
+	}
+}
+
+// Register Appliance Management API paths with the vapi simulator's http.ServeMux
+func (h *Handler) Register(s *simulator.Service, r *simulator.Registry) {
+	s.HandleFunc(consolecli.Path, h.consoleCLIStatus)
+	s.HandleFunc(dcui.Path, h.dcuiStatus)
+	s.HandleFunc(ssh.Path, h.sshStatus)
+	s.HandleFunc(shell.Path, h.shellStatus)
+}
+
+func (h *Handler) decode(r *http.Request, w http.ResponseWriter, val interface{}) bool {
+	return Decode(r, w, val)
+}
+
+// Decode the request Body into val.
+// Returns true on success, otherwise false and sends the http.StatusInternalServerError response.
+func Decode(request *http.Request, writer http.ResponseWriter, val interface{}) bool {
```

Comment:
my opinion: since you return true/false you should not set the header at all in this func since

a) it won't allow the caller to write a specific header (ie not generic function)
b) the caller likely won't do anyhting with the return parameter since you already set header here - being more explicit on the caller side is better IMHO, e.g. inspec true/false and write header accordingly from caller

Reply:
Updated the code to update the header on the caller side based on the return value.
```
if h.decode(request, writer, &input) {
  .....
  writer.WriteHeader(http.StatusNoContent)
} else{ 
writer.WriteHeader(http.StatusInternalServerError)
}
```